### PR TITLE
change sm provided credentials level

### PR DESCRIPTION
**Pull request by sigalmaya**
move sm_provided_tls_credentials to tls object

### CODE REVIEW

File: pkg/types/credentials.go [hunk 1]
```diff
@@ -146,7 +142,7 @@ func GenerateCredentials() (*Credentials, error) {
 }
 
 func (c *Credentials) TLSExists() bool {
-	return c.TLS != nil && *c.TLS != TLS{}
+	return c.TLS != nil && !(c.TLS.Certificate == "" && c.TLS.Key == "")
```
Comment (pavelmaliy) [inline pkg/types/credentials.go:145]: change it to CertificateExists

File: pkg/types/credentials.go [hunk 2]
```diff
@@ -50,16 +50,13 @@ func (c *Credentials) MarshalJSON() ([]byte, error) {
 	if toMarshal.Basic == nil || toMarshal.Basic.Username == "" || toMarshal.Basic.Password == "" {
 		toMarshal.Basic = nil
 	}
-	if toMarshal.TLS == nil || toMarshal.TLS.Certificate == "" || toMarshal.TLS.Key == "" {
-		toMarshal.TLS = nil
-	}
```
Comment (pavelmaliy) [inline pkg/types/credentials.go:53]: why you removed it?

Comment (pavelmaliy) [reply pkg/types/credentials.go:53]: maybe like this?
if toMarshal.TLS == nil || (!UseSMCertificate && (toMarshal.TLS.Certificate == "" || toMarshal.TLS.Key == "" )) {
              toMarshal.TLS = nil
}

**Review by pavelmaliy — DISMISSED** [review]

File: pkg/types/credentials.go (hunk 1 above)
Comment (evyaffe) [reply pkg/types/credentials.go:145]: !(c.TLS.Certificate == "" && c.TLS.Key == "") --> (c.TLS.Certificate != "" || c.TLS.Key != "")
Shouldn't it be (c.TLS.Certificate != "" && c.TLS.Key != "")?
Unless I got something wrong here...

Comment (pavelmaliy) [reply pkg/types/credentials.go:145]: you right they both can't be empty, good catch
c.TLS.Certificate != "" && c.TLS.Key != ""

File: pkg/types/credentials.go (hunk 2 above)
Comment (sigalmaya) [reply pkg/types/credentials.go:53]: but it is useless,  because we ahve also "omitmpty " on fields

File: pkg/types/credentials.go (hunk 1 above)
Comment (sigalmaya) [reply pkg/types/credentials.go:145]: Why it is wrong? we did it on purpose(for the update case) : if c.CertificateExists() {
		err := c.validateTLS() 
and then in validate tls we check each field 

func (c *Credentials) validateTLS() error {
	if c.TLS.Certificate == "" || c.TLS.Key == "" {
		return errors.New("tls public certificate and key should be provided")
	}

	if c.TLS.UseSMCertificate {
		return errors.New("only one of the options could be set, SM provided credentials or tls")
	}
	_, err := tls.X509KeyPair([]byte(c.TLS.Certificate), []byte(c.TLS.Key))
	if err != nil {
		return errors.New("invalidate TLS configuration: " + err.Error())
	}

	return nil
}

File: pkg/types/credentials.go [hunk 3]
```diff
@@ -32,16 +32,16 @@ type Basic struct {
 }
 
 type TLS struct {
-	Certificate string `json:"client_certificate,omitempty"`
-	Key         string `json:"client_key,omitempty"`
+	Certificate      string `json:"client_certificate,omitempty"`
+	Key              string `json:"client_key,omitempty"`
+	UseSMCertificate bool   `json:"sm_provided_tls_credentials,omitempty"`
```
Comment (evyaffe) [inline pkg/types/credentials.go:37]: the naming here is a bit odd, why not leave it SMProvidedTLSCredentials?
later in broker ToObject you use 'tls := &types.TLS{UseSMCertificate: e.SMProvidedCredentials}', so you need to match the strings, either change in the broker object or keep this one as is. 

WDYT?

Comment (sigalmaya) [reply pkg/types/credentials.go:37]: Eyal  told us  to use UseSMCertificate after we already agreed on API with xrs.  What do you think is better to use "useSMCerificate" or "SMProvidedTLSCredentials" i will align

File: pkg/types/credentials.go (hunk 1 above)
Comment (pavelmaliy) [reply pkg/types/credentials.go:145]: you right I forgot it followed by validate call, 
you wanted to assure here that update leaves inconsistent data ignore my comment

Comment (evyaffe) [reply pkg/types/credentials.go:145]: agree, resolving

File: pkg/types/credentials.go (hunk 2 above)
Comment (sigalmaya) [reply pkg/types/credentials.go:53]: it is used by JSON library.

File: pkg/types/credentials.go (hunk 3 above)
Comment (pavelmaliy) [reply pkg/types/credentials.go:37]: please change it to SMProvidedTLSCredentials it's confusing.

**Pull request merged**